core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -89,9 +88,8 @@ public void init() {
                   Boolean expired = CompletionStages.join(expiredStage);
                   handleExpiredEntryContextAddition(expired, ctx, readEntry, key, isOwner);
```

Comment:
shouldn't wait for `previousStage` before invoking `handleExpiredEntryContextAddition()`?

Reply:
No, it's a bit more complicated than that... It's ok to access the invocation context from the same thread, OR from another thread, after `previousStage` is complete.

The javadoc says

```
    * @param previousStage don't access the invocation context from another thread before this stage is complete
```

I'll try to reword a bit to make the key words "from another thread" do less work :)

Reply:
actually, I'm wondering if the "fix" wouldn't be simply to replace the `HashMap` with a `ConcurrentHasHMap` in `NonTxInvocationContext`